deliaioanac/assignment2/LCA.cpp

```diff
@@ -0,0 +1,157 @@
+#include <iostream>
+#include <vector>
+#include <math.h>
+#include <stdio.h>
+#include <utility>
+#include <stdlib.h>
+
+
+// Structure for a tree node
+struct TreeNode {
+    int val;
+    std::vector<TreeNode*> children;
+    TreeNode(int x) : val(x) {}
+    ~TreeNode() {
+        while (!children.empty()) {
```

Comment:
[optional] If you have some time, you should look into [std::unique_ptr](http://www.drdobbs.com/cpp/c11-uniqueptr/240002708).

It's a wrapper over raw pointers that makes memory management super simple. Basically if you wrap everything in a unique_ptr, the object will be deleted automatically without needing to specify destructors/ calling delete on objects you've created. It's the preferred way in C++.

Reply:
I actually did a short training on all the innovations brought by C++11 and 14 two years ago, like smart pointers, move and copy constructors, universal references etc, but it was all theoretical and, while they sounded amazing (automatic destruction, less memory overhead), I never actually got around to trying to use them in my code. I will modify my code to use them. I apologize in advance for any mistakes, there might be a few, but I'll do my best to learn from them.

Reply:
Your code looks good! The added benefit of smart pointers is if you're using them wrong, the compiler will be quick to tell you ;)

Don't worry about not knowing all of the latest C++ stuff, C++ is a complex language. It takes a really long time to master it. The only reason I'm even pointing out all that stuff is because your code is already pretty good otherwise :)

Reply:
I appreciate that you take the time to tell me all this, I am learning so much!